python/sqlflow_submitter/tensorflow/input_fn.py

```diff
@@ -117,43 +108,34 @@ def input_fn(select,
     return dataset.map(ds_mapper)
 
 
-def pai_maxcompute_db_generator(table,
-                                feature_column_names,
-                                label_column_name,
-                                feature_specs,
-                                fetch_size=128,
-                                slice_id=0,
-                                slice_count=1):
-    def reader():
-        selected_cols = copy.copy(feature_column_names)
-        if label_column_name:
-            selected_cols.append(label_column_name)
-            try:
-                label_idx = selected_cols.index(label_column_name)
-            except ValueError:
-                # NOTE(typhoonzero): For clustering model, label_column_name may not in field_names when predicting.
-                label_idx = None
-        else:
-            label_idx = None
-        reader = paiio.TableReader(table,
-                                   selected_cols=",".join(selected_cols),
-                                   slice_id=slice_id,
-                                   slice_count=slice_count)
-        while True:
-            try:
-                row = reader.read(num_records=1)[0]
-                label = row[label_idx] if label_idx is not None else -1
-                features = []
-                for name in feature_column_names:
-                    feature = read_feature(row[selected_cols.index(name)],
-                                           feature_specs[name], name)
-                    features.append(feature)
-                if label_column_name:
-                    yield tuple(features), label
-                else:
-                    yield (tuple(features), )
-            except Exception as e:
-                reader.close()
-                break
-
-    return reader
+def parse_pai_dataset(feature_column_names, has_label, feature_specs, *row):
+    features = {}
+    for i, name in enumerate(feature_column_names):
+        spec = feature_specs[name]
+        f = read_feature(row[i], spec, name)
+        features[name] = tf.SparseTensor(*f) if spec["is_sparse"] else list(f)
+    return features, row[-1] if has_label else features
+
+
+def pai_dataset(table,
```

Comment:
Please take a look at https://github.com/sql-machine-learning/sqlflow/pull/1878/files#diff-f1dc5fd65c894da64814798d3954cebaL116, the parsing function need to deal with many data input formats.

Reply:
The `read_feature` function is handling those input formats. Any omission？

Reply:
I see. That should be correct, I'll test this PR using the internal CI before merging.